static/pane.ts

```diff
@@ -0,0 +1,195 @@
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import _ from 'underscore';
+import { Container } from 'golden-layout';
+import * as monaco from 'monaco-editor';
+
+import { FontScale } from './fontscale'
+import { BasePaneState, PaneCompilerState } from './pane.interfaces';
+import { SiteSettings } from './settings.interfaces';
+
+/**
+ * Basic container for a tool pane in Compiler Explorer
+ *
+ * Type parameter E indicates which monaco editor kind this pane hosts. Common
+ * values are monaco.editor.IDiffEditor and monaco.ICodeEditor
+ */
+export abstract class Pane<E extends monaco.editor.IEditor> {
+    compilerInfo: PaneCompilerState;
+    container: Container;
+    domRoot: JQuery;
+    topBar: JQuery;
+    eventHub: any /* typeof hub.createEventHub() */;
+    selection: monaco.Selection;
+    editor: E;
+    fontScale: typeof FontScale
+    isAwaitingInitialResults: boolean = false;
+    settings: SiteSettings | {} = {};
+
+    /**
+     * Base constructor for any pane. Performs common initialization tasks such
+     * as registering standard event listeners and lifecycle handlers.
+     *
+     * Overridable for implementors
+     */
+    protected constructor(hub: any /* Hub */, container: Container, state: BasePaneState) {
+        this.container = container;
+        this.eventHub = hub.createEventHub();
+        this.domRoot = container.getElement();
+        this.initializeDOMRoot();
+
+        const editorRoot = this.domRoot.find('.monaco-placeholder')[0];
+        this.createEditor(editorRoot);
+
+        this.selection = state.selection
+        this.compilerInfo = {
+            compilerId: state.id,
+            compilerName: state.compilerName,
+            editorId: state.editorid,
+        }
+        this.fontScale = new FontScale(this.domRoot, state, this.editor);
+        this.topBar = this.domRoot.find('.top-bar');
+
+        this.registerButtons(state);
+        this.registerStandardCallbacks();
+        this.registerCallbacks();
+        this.setTitle();
+    }
+
+    /**
+     * Initialize the DOM root node with the pane's default content. Typical
+     * implementation looks like this:
+     *
+     * ```ts
+     * this.domRoot.html($('#rustmir').html());
+     * ```
+     */
+    abstract initializeDOMRoot(): void;
+
+    /**
+     * Initialize the monaco editor instance. Typical implementation for looks
+     * like this:
+     *
+     * ```ts
+     * this.editor = monaco.editor.create(editorRoot, extendConfig({
+     *     // goodies
+     * }));
+     * ```
+     */
+    abstract createEditor(editorRoot: HTMLElement): void;
```

Comment:
Different panes use different monaco editor kinds and monaco editor options, so delegating the editor creation to a method makes more sense for implementors.